Added code which provides callbacks from external ML-IAP functions to…

bathmatt opened the pull request:
… exchange data

This is designed for message passing ML models.  Currently FP32/64 values can be communicated but these are passed via doubles in MPI,
Work completed by
Forrest Glines <fglines@nvidia.com>
Matt Bettencourt <mattb@nvidia.com>

**Summary**

<!--Briefly describe the new feature(s), enhancement(s), or bugfix(es) included in this pull request.-->

**Related Issue(s)**

<!--If this addresses an open GitHub issue for this project, please mention the issue number here, and describe the relation. Use the phrases `fixes #221` or `closes #135`, when you want an issue to be automatically closed when the pull request is merged-->

**Author(s)**

<!--Please state name and affiliation of the author or authors that should be credited with the changes in this pull request. If this pull request adds new files to the distribution, please also provide a suitable "long-lived" e-mail address (ideally something that can outlive your institution's e-mail, in case you change jobs) for the *corresponding* author, i.e. the person the LAMMPS developers can contact directly with questions and requests related to maintenance and support of this contributed code.-->

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

<!--Please state whether any changes in the pull request will break backward compatibility for inputs, and - if yes - explain what has been changed and why-->

**Implementation Notes**

<!--Provide any relevant details about how the changes are implemented, how correctness was verified, how other features - if any - in LAMMPS are affected-->

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

fglines-nv committed:
Added code which provides callbacks from external ML-IAP functions to exchange data
This is designed for message passing ML models.  Currently FP32/64 values can be communicated
but these are passed via doubles in MPI,
Work completed by
Forrest Glines <fglines@nvidia.com>
Matt Bettencourt <mattb@nvidia.com>

bathmatt commented:
@fglines-nv @bnebgen-LANL 
Here is the kokkos/lammps side of the changes required to get message passing into hippy-nn.  There are changes to that code base which we're going through the release process.

Commit:
fixed whitespace

akohlmey commented:
@bathmatt FYI, if you submit an incomplete PR, you should make it a "Draft pull request" and then click on the "Ready for review" button after it is complete.

rbberger commented on src/KOKKOS/pair_mliap_kokkos.cpp:
```diff
@@ -113,7 +116,9 @@ void PairMLIAPKokkos<DeviceType>::compute(int eflag, int vflag)
-
+  int nrank,size;
+  MPI_Comm_rank(MPI_COMM_WORLD,&nrank);
+  MPI_Comm_size(MPI_COMM_WORLD,&size);
```
Unused code.

Commit:
Remove dead code

Commit:
Changed use of varient on LHS

bathmatt committed:
Merge branch 'develop' into lammps-kokkos-mliap-comms

bathmatt commented:
Ok, this is ready to review, not sure who. I just updated the branch,

fglines-nv committed:
Make Pytorch optional

fglines-nv committed:
Clean up cupy vs. torch in exchanges

fglines-nv committed:
Fix tabs->spaces in last commit

bathmatt committed:
Merge branch 'develop' into lammps-kokkos-mliap-comms

bathmatt commented:
@rbberger I was waiting for functionality review by @lubbersnick or another person from LANL, but we have a group using it here at NVIDIA, different from ours, so we feel it is ready to merge into master.  Are you the official PR review person for LAMMPS?  This was for HIPPY-NN could do message passing in its ML model. (We're playing with other message passing ML models at the current moment in LAMMPS as well)

lubbersnick commented:
I wasn't directly aware of this PR. The biggest thing to me is that it looks like this feature only exists using ML-IAP-kokkos, so it is not compatible with "vanilla" ML-IAP. I don't know if that's worry for the core developers, maybe @athomps could weigh in- does one need feature parity between kokkos and non-kokkos?

rbberger commented:
@bathmatt yeah, I've assigned myself as first reviewer after our last core dev meeting (I'm also LANL) and was waiting for the Draft status to disappear. You also already seem to have addressed one of the concerns previously raised by other core devs about the PyTorch dependency. 

@lubbersnick We normally don't force feature parity between kokkos and non-kokkos parts, but we still have to verify that optimizations like this don't break any existing code. I'll also have to look at existing examples, since I've been told that simple LJ tests started to fail some time ago (unrelated to this PR). I'll get to it in the coming days.

fglines-nv commented:
@rbberger The additions shouldn't affect existing code since the optimizations are opt-in. One has to call these new hooks to see any differences. Still worth checking existing code, however

bathmatt commented:
Do we need to keep the backwards compat to the non kokkos side, please let me know and I can help produce that.

rbberger commented:
Finally got to test this. LGTM.

I also figured out what others meant with the "PyTorch tests being broken". This has nothing to do with this PR.

---
Just to be clear, I'm not asking you to fix this here, but I'd appreciate if someone from this group can take a closer look and fix these issues. Moving the description about it to https://github.com/lammps/lammps/issues/4526

bathmatt committed:
Merge branch 'develop' into lammps-kokkos-mliap-comms

bathmatt commented:
I addressed the comments in the other thread about #4526 and rebaselined this MR, it is suggesting @stanmoore1 to also review and maybe push the button.

stanmoore1 committed:
whitespace
